# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Identify antibody-derived-tags from single cell RNAseq data
#' 
#' @param bamFileName The input BAM/SAM filepath, this is typically output from the dropseq-tools pipeline.
#' @param bamOut The output BAM/SAM filepath
#' @param adtFasta Filepath for the fasta file containing the TotalSeq antibody barcode sequences (in fasta format)
#' @param max_dist integer. The maximum Hamming distance allowed for a barcode match to be called
#' @param sumoutput Filepath for tab delimited text file output of cell-barcodes, umi, antibody, hamming distance and map score for each record in the output bam/sam
#' @export
adtseq <- function(bamFileName, bamOut, adtFasta, max_dist, sumoutput) {
    .Call('_adtseq_adtseq', PACKAGE = 'adtseq', bamFileName, bamOut, adtFasta, max_dist, sumoutput)
}

#' Extraction of read names from specific cell barcodes.
#' 
#' @param bamIn The input BAM/SAM filepath, this is typically output from the dropseq-tools pipeline (string).
#' @param bamOut The output BAM/SAM filepath (string).
#' @param sumoutput Filepath for tab delimited text file output of cell-barcodes, umi, fastq read name (string).
#' @param cellBC Character string specifying the 12 nt cell barcode for which you want to extract reads for.
#' @export
bcExtract <- function(bamIn, bamOut, sumoutput, cellBC) {
    .Call('_adtseq_bcExtract', PACKAGE = 'adtseq', bamIn, bamOut, sumoutput, cellBC)
}

#' Pairwise Hamming Distance
#' 
#' @param seqs Character vector of equal length strings containing the sequences from which to calculate pairwise Hamming/String distances.
#' @export
hdist <- function(seqs) {
    .Call('_adtseq_hdist', PACKAGE = 'adtseq', seqs)
}

